### DISCUSSION TITLE
[Feedback request] What is your usage of Smart AutoClicker ?

### DISCUSSION THREAD

**Discussion by Nain57** [Nain57/Smart-AutoClicker#57] [General]
Hi everybody !

As you probably know, Smart AutoClicker is doing zero data collection in order to protect the user privacy. This means, I'm completely unaware of your usage of this application. So, in order to improve Smart AutoClicker for the general usage, it would be nice if you could share with us your experience. 

- How many scenarios do you have ?
- How many events in average ?
- What are you automating ? A game (which one) ? something else ?
- I don't need to know everything, but anything you think might be interesting  for me can be great !

Thank you !

**Comment by sulivanShu** [comment]
Hello, I wish use your program for activate and desactivate Google ads and Microsoft ads campagns on my smartphone. I have 6 campaigns each requiring 3 or 4 clicks.

For the moment, I am discovering the application. At the moment, I have some difficulties, although I managed to program a few clicks.

**Reply by Metazolid** [reply]
If you want to remove ads, I think it's better to use AdAway, which is also available on F-Droid.

**Comment by ghost** [comment]
I use it for game login and auto-farming.  It would be nice if there would be a randomization component to the swipe zones.  It took me awhile to figure out how to get the image detection running too, and I had to fiddle around with the wait times.  A lot of these types of apps use lua scripting so if it could produce autogenerated code and import scripts via the UI that would be fabulous.  There are open source projects to this effect, so it shouldn't be overly difficult to implement and understand what's going on.  Take a look at easy macro maker and the another eden fish bot, which uses fuzzy image recognition and automatically moves your screen image around for a best match, although this relies on windows in Windows applications.  Products like Ankalua and Macrorify are charging obscene amounts for simple scripting that's readily available on iOS and various Windows applications so it's fantastic that this exists and is open sourced.

I strongly suggest that there be sample videos on YouTube to introduce basic use cases, because there is no useful help or documentation whatsoever.

**Comment by irwanwr** [comment]
For games mostly for now. Especially on games with a lot of grinding. 
I wish this app can recognise and click on recognised image dynamically. Not just on fixed spot or area of the screen. 
Very useful app indeed. Thanks to developer.

**Comment by metroxx** [comment]
So well. Currently i'm trying to make a leech script for World flipper. So i could get over the multiplayer grinding quite fast. Because fights takes from seconds to 10-15 minutes. My only problem is that i can't tell it to do specific action once. Or i just haven't figured it out.

**Reply by Nain57** [reply]
It's not possible at the moment, but this feature has been requested many times and should be implemented soon.

**Comment by sudomain** [comment]
I hesitated to join this discussion because the truth is I don't use this app that often and I didn't want to negatively affect the motivation of the developer. I mostly defer to scripting things in Termux for portability (Bash, Python) or using [Key Mapper](https://github.com/sds100/KeyMapper) which takes a different approach to automation. Despite the name, Key Mapper's actions can be triggered from other apps such as Termux's version of `am` (Android activity manager). It also allows for constraints such as only performing actions when certain apps are or aren't in the foreground (possibly a good feature for Smart AutoClicker?) which is something I need. 

I didn't intend post here to advertise for the "competition" of Smart AutoClicker: Sorry if it came off that way. These are just the tools I use and how I generally use them. KeyMapper is currently limited in its screen tapping capabilities with just an action for `single tap at X, Y coordinate`, though to be fair I don't think anyone has made feature requests for swipes or image recognition. Smart AutoClicker does these things very well and I think an intent receiver to play scenarios (#44) would greatly help users integrate it into their existing automation setups. 

To get to the main topic: There are actions that would normally require root or that are difficult in other FOSS automation apps. Smart AutoClicker, due to it's image recognition, could do these things very well:

1. Intelligently toggling mobile data (root)
2. Rebooting the device (root)
3. Automatically switching to a certain keyboard/input method (root)
4. Checking to see if part of web page has changed by copying/pasting a web page (that I can't `wget` due to a login requirement) into a Termux script running `diff`
5. Clicking checkboxes of a daily health screener if the above found no differences
6. Selecting appropriate credentials in KeePassDX or another password manager with autofill (Smart AutoClicker users should not use it to copy/paste passwords because the Android  clipboard is known to be insecure. High value credentials probably shouldn't be on an android device to begin with).
7. Games, as others have mentioned, but also web apps that can only be accessed via the browser.

EDIT: Thank you for the app. I reread my first paragraph and want to be very clear that I'm grateful for this app and there's clearly many things Smart AutoClicker can do that other FOSS/FDroid apps can't.

TODO: feature requests I thought of while writing this long post:
Import/export scenarios #81 
Send intent action #82
Select/context menu tap action

**Comment by Metazolid** [comment]
I've been using your tool _a lot_.
Pretty much only for games, where I could let my phone grind troughout the night, sometimes leaving it in the fridge lol.
I use 1.0 RC2 on my phone (at least I think I do, it's a slightly older version) and the most recent one on Bluestacks.
The reason why I haven't upgraded is that I like to keep my scan areas as small as I can, whereas the most recent version only allows for a larger area which can/will include undesirable UI that can trigger false positives and while the waiting times and more accurate swipes are 100% an improvement, I haven't been in a situation where I needed them so far.

I have 8 scenarios on my phone, but only use 2-3, which have 6-9 clicks.
1 on Bluestacks with 5 events.

The Tower Game and C.A.T.S. are the ones I have primarily tinkered with.

**Reply by ilovestowerin** [reply]
Not sure if you check your GitHub, but I am curious about if you've been detected on the tower game since using?

**Comment by Tsyoee** [comment]


**Reply by Nain57** [reply]
Hi,

Do you encounter any issue with the current version (1.3.1) ? If yes, please open a ticket in the [issue tracker,](https://github.com/Nain57/Smart-AutoClicker/issues) I will gladly help :)

**Reply by fogoat** [reply]
BlueStacks X now has enhanced macro with "scripting". It supports loops, etc. Might be useful if SmartAutoClicker doesn't do it all for you.

**Comment by TWT233** [comment]
How many scenarios do you have ?

- 2, but for the same app

How many events in average ?

- 4

What are you automating ? A game (which one) ? something else ?

- game, arknights

I don't need to know everything, but anything you think might be interesting for me can be great !

- yea this app is a really convenience for some dumb farming gacha games, thanks for saving my life in the first
- feat req: wanna more precise condition combination control: 
  - now there is only "all" and "oneof" strategy for condition combination
  - it's useful to grant a mechanism like boolean condition combination imo., such as: (cond1 & cond2) || cond3

**Reply by Nain57** [reply]
Hi ! Thanks for your feedback.
For the combination controls: at first, I was aiming at the same kind possibilities, but it turned out to be very complicated to find a simple UI for it. 
It is good to know that you can achieve similar results by creating a copy of your event and changing the conditions on it:

Let's say you want, with your method:

Event1 [(cond1 & cond2) || cond3)]


With my method, you will have a similar result by doing:

Event1 [(cond1 & cond2)]
Event1Copy [cond3]

Hope this helps !

**Comment by TWT233** [comment]
Thanks for your note, that’s a inspiration.

Since the UI is the matter which I have no idea about,
may another feat req: negative condition(the current screen not match the target detection -> true)?

On Apr 25, 2022, at 16:39, Nain57 ***@***.******@***.***>> wrote:



Hi ! Thanks for your feedback.
For the combination controls: at first, I was aiming at the same kind possibilities, but it turned out to be very complicated to find a simple UI for it.
It is good to know that you can achieve similar results by creating a copy of your event and changing the conditions on it:

Let's say you want, with your method:

Event1 [(cond1 & cond2) || cond3)]

With my method, you will have a similar result by doing:

Event1 [(cond1 & cond2)]
Event1Copy [cond3]

Hope this helps !

**Comment by rejfin003** [comment]
Im automating a game where I can click a health potion button repeatedly until the health level is fully restored.

Could we have two set of actions? One for negative and one for positive result of the conditions.  Image below is just for reference. 

![image](https://user-images.githubusercontent.com/110021039/180959885-5d1fa87b-2d07-4184-9d1b-280e58d48726.png)

**Comment by ordipokia** [comment]
Hi !
I am French, so I use a known translator to put the message in English. Excuse me if the translation is not correct.
Thank you for this very easy to use app.
I use actiona on windows and it took me a while to find an app on android.
I am using version 1.3.1
     * How many scenarios do you have?
10 on multiple phones
     * How many events on average?
from 3 to 8
     * What are you automating? A game (which one)? something else ?
Crypto games and mining
it is also used to close the ads (which must be displayed to validate the action so I cannot use ad blockers)
     * I don't need to know everything, but whatever you think is interesting to me can be awesome!
I haven't looked at version 1.5 and especially the new UI
On version 1.3.1:
 - From time to time the size of the buttons bother me on the screen.
 - I reduce the area of ​​the detection image each time. could it not be saved for the next detection?
- in a scenario call another scenario. I have to copy my ads closing scenario in each of the other scenarios which is complicated to add a new detection.
- launch phone actions. Like HOME or BACK
- possibly have 2 scenarios simultaneously. the objective being that on one there is a timer which closes all the applications (HOME) and restarts one (even by image detection. this is what I use to launch another application)


Thanks very mutch for this app !

**Comment by RaufSab** [comment]
I came across your app while Iwas looking for an auto clicker which can detect the location of a customers or the zip code of the area of which the restaurant orders coming from of the food delivery app I'm using.

It will help me a lot if you ever developed a logarithm which can work well for this feature.

Best regards.

**Comment by anwbeer** [comment]
I have carpal tunnel and the less I have to tap on the screen, the better it is for my health.

I was looking for a macro that could do an action based on image recognition. I use it click on Google reCaptcha but the prompt could be in 5 minutes or 5 hours. Performing the action based on the image is more reliable in this case. 

One request I'd like to ask is better detection of ad close buttons (usually a white circle with a x inside). Maybe changing the screen image to B&W, or somehow filtering the background image, so certain ad shapes like white with a X or a double sideways chevron (to go to Play Store) are detected reliably and the correct action performed.

**Comment by shmaungms** [comment]
Hi,

I use auto clicker with 2 different setups with a game I play.
1. Firstly to navigate some menus/screens. I have to do a particular battle mode 30 times a day in the game but from the end of the battle to being able to start the next takes about 10 - 12 seconds of clicking through screens. So I use auto clicker 2 automatically navigate the menus/screens back to the battle again which requires three different detection and three different clicks on the screen. Given each battle also only takes 10 seconds and only a glace or 2 at the screen I don't need to pay attention to the game and I often read the news or something whilst completing these battles.
2. The game also relies on XP to build up your strength. Being a free to play game you need to do many, many battles in a different battle mode than the one above to gain decent XP. To help this along to game does provide an auto mode to play the battles. I therefore use this auto clicker to click on and start the auto play mode then once the battle is finished to navigate multiple screens and menus to start the next battle and then start the auto battle mode again.  Because you get rewards from completing the battles and there is a story attached to the battles to perform this loop I actually have nine different events set up.  To navigate these screens manually it takes 15 - 30 seconds depending if there is story or chapters or not. Most with just one click and one condition, but some have multiple conditions (only needs one to trigger) and some have multiple clicks in a single event.

In my scenario with this particular game this smart clicker does everything I could want. Some other auto clickers out there offer anti-detection settings in their options. I'm assuming this varies the time between clicks and also the exact location of the click on the screen. I'm not sure how many games/apps actively try to detect clickers or not. Not sure if you can add something like this to a future version to be on the safe side for your users.

And if you could do anything to reduce the power consumption of this app that would be appreciated too ;) Running just the game while plugged into a standard USB computer port the phone battery charge percentage would slowly go up but with your app running as well as the game the battery percentage slowly goes down under the same conditions.

Hope this helps.

**Comment by invidtiv** [comment]
I use it inn a game , I have around 11 scenarios, 4 are still in dev.
I use them a lot, the backup and restore function allows me to sync the latest version of any scenario with any of the emulators I use.
For me one add on essential is the possibility to take a screenshot.
A great feature would be to be able to turn of scenarios with conditions and turn them back on with other conditions, so like have an action to disable or enable conditions within a scenario.
Another great feature would be to have variables and being able to use then in conditions.
A dream feature would be to have OCR being able to place values in variables and being able to compare those variables to create conditions like greater than.

bugs : quite a few issues of the game killing the app because of low memory

Also a few examples with the intent, in a doc, would be greatly appreciated.

Also thank you for the great work

**Comment by lautit** [comment]
Hi! I've been using this app a lot lately. Many thanks for developing and open sourcing it to the community, you rock!

The usage has been mostly with the game Summoner's Greed. It involves a lot of AFK farming with recurring events, which are easily detectable and I've been able to make 3 scenarios to auto farm these levels. They have between 5 to 9 events each.

The game also has some exploity mechanics with the progress saving on Facebook, which lets you re-roll multiple times without wasting resources or time. I've been able to orchestrate the FB Connect/Disconnect once and the copy event feature has been life saver for the 3 other scenarios using this exploit. Due to higher complexity, this scenarios have between 6 to 15 events each, and here is where I would like to suggest some features.

In these complex scenarios, there are a couple of repetitive patterns I've been able to perform to produce a flow of validation:

- Disable self: using the event toggling, I disable the same event on triggering in order to just fire once. This would be amazing to either include as another event or a checkbox in the event itself.
- Multiple toggling: using the event toggling, I also created some dummy events which only enable/disable multiple other events for easier connection. It would be nice to have either a mux for toggling other events, or even some starting/ending procedures for events.
- Present/absent conditions: through multiple events, I've been able to catch either presence or absence of the condition, effectively routing the procedure on the right path. It would be nice to include a deviation in case of condition found or not, this could be configured per event too.

Also, there are a couple of QoL improvements you may find interesting:

- Copying scenarios
- Procedure of events: a procedure of events, that can be triggered on a condition. This could be simply using scenarios as events, for example Would greatly improve repetition.
- Custom events: a list of common events that could be reused across scenarios. Would greatly improve modifications across multiple scenarios of copied events.
- Re-cutting conditions: once you take and crop a screenshot, you can only use it as-is on multiple events. If the full screenshot is saved, it would greatly improve reusage on cropping other areas, or fixing bad cropping, since some screenshots are difficult to re-take due to probabilities on some games.
- Timeout on N event triggers: a new scenario end condition to, instead of fully stopping the scenario, just "let it sit" waiting for a timeout to ocurr and re-starting. Useful for waiting for FB Login API refresh and avoid shadowbans.

I think these are some of the simpler improvements, that would indeed make the UI more complex but the tool more powerful. In order to avoid user confusion, any of these functionalities can be enabled by some toggles on the main app menu, so to let advanced users play a little bit more.

As an endgame, I believe the UI would be much more powerful if the events and scenarios could be managed in a node-like graph, linking stuff between each other, effectively building a state machine of some sort. You can find some inspiration on the [while True: learn()](https://store.steampowered.com/app/619150/while_True_learn/) game, I found it both entertaining and inspiring!

BTW, if you need an extra hand, I've been trying to dive into the code to implement some of these features myself. We can connect and do some onboarding for easier understanding of the codebase and do some pair programming! I would be more than glad to do so.

Again, infinite thanks for making such an amazing and entertaining app, and open sourcing it as the icing on the cake. I adhere to the donation links in order to retribute a little bit of your time.

**Comment by just-shadyumbrella** [comment]
It would be nice if Klick'r had advanced scripting feature or macro like [Macrorify](https://play.google.com/store/apps/details?id=com.kok_emm.mobile) but has also easy to use flow mapping like [Automate](https://play.google.com/store/apps/details?id=com.llamalab.automate), Klick'r is already good enough for decent task but maybe lack for advanced (like [randomization](https://github.com/Nain57/Smart-AutoClicker/discussions/57#discussioncomment-1526407)), and also good suggestion as [mentioned](https://github.com/Nain57/Smart-AutoClicker/discussions/57#discussioncomment-7269570).

And also if it isn't good enough, a text or android view recognition (like control, button, switch, etc) using Accessibility API directly rather than image recognition (as last resort).

I had an issue on MIUI based device that saying Klick'r is using Accessibility but there's nothing happened on the screen (script not working), I guess that's from the broken MIUI itself. But after all, it's been already good enough for certain automation task, but I hopes it's can be more advanced as mentioned.

**Comment by willross42** [comment]
Great app! Already helped me with the grind in a couple of games, so that I can focus on more fun parts in the less and less time I have to play games as life goes on. 

It would be good to be able to create subroutines that can be added to different scenarios, with the ability to set the number of times to repeat the subroutine.  For example, I am playing a game where I want to start a process through a dialogue with an NPC then navigate to somewhere nearby with a good amount of space,  run back and forth a certain number of times until that process is complete, then finally navigate back to the NPC.  Creating 100 separate swipes just for the running back and forth bit is very cumbersome and messy when it could be two swipes inside a subroutine that is repeated 50 times. Also, I intend to build this into a larger smart routine and it would be good to be able to reuse chunks for the simple scenario built with a  few subroutines.

**Reply by Deafspeaker** [reply]
I have been using on a game. I've gotten the image detection working for what I need. A couple of the aspects of the game are harder, 1 because the screen slowly rotated around on its own to prevent botting, I get it. But that means I would need to take so many detection images on some of the nodes. Then there is some things that require the player to actually move a step before gathering the node again. Trying to figure out how to get past that. The biggest one is the fishing. The fish node creates a random group o lf circles that grow and fade and the pattern is always random so the detection doesn't seem to get it.. tricky..

**Comment by MarkR2030** [comment]
I've used Klick'r since before the name change (Great program, btw!), mostly for grinding in a handful of games and sometimes auto-skipping their ads. 
Recently, since another independent program that I used, Cygery AdSkip for YouTube (XDA Developers,  closed-source, no longer active), finally got completely borked by Google Ad Services, I began trying to mimic its tag- and logcat-based functions with this pgm's unusually well-developed visual recognition feature. 
Much like klick'r, it would detect and mute yt ad audio, then automatically unmute and press [Skip] or [Next]. It came in handy when I'd be forced to use the official ewwtube app while Google made web API <ahem> *adjustments*, or just casually browsing without keeping up with that whole Google vs Hackers, war-of-attrition ping pong dynamic.
I say, "trying", because I'm having difficulty getting the screen captures, their background variations, and another problem with mute/unmute intents glitching(staying stuck on or off by misreading system audio state for toggle).
My success rate is still well below 50%, but I really just got started. 
In fact, I just restarted completely from scratch, 
<<all while wishing this pgm had an unofficial repository of scenarios and guides for advanced users and features.>>

**Comment by GrimreeprSask** [comment]
I use it for several "idle" games on my LG Velvet and Samsung s22.  The latest version however has several issues and refuses to work.  Detections are not firing, and when they do, the action chosen for the detection wont trigger.  Sometimes everything triggers properly, then the next time, it wont work.  It's become unusable for my purposes.

**Comment by zianifique** [comment]
First of all, thank you so much for developing such an app!

I use auto-clickers mainly on my Android phone, to automate grinding on a mobile game called The Tower. I discovered Klick’r a couple of days ago. Until then, I was using ‘blind’ auto-clickers, figuring out tricks to have stable loops.

I already developed two Scenarios, each with about 15 Events/Triggers. The image recognition works very well, after spending a little time adjusting the extension of the image to recognize and the error threshold.

I could not figure out how to use Intent actions to take and save screenshots on my device, so I tried combining Klick’r with another app meant to take screenshots. Unfortunately, it did not work, probably because both these two apps have competing permissions to stay over each other. So my only request would be about adding a way to save screenshots on the device.

Once again, I really love this app. It is very well thought through in its structure, and its tutorial is very well designed.

**Comment by colineberly** [comment]
Amazing app. I use it on emulators for gaming every day. It automates like 80% so that I can focus on other things while basically farming. It's such an intelligent app, I really feel like I should be using it for other things, but I haven't really dove into it quite yet other than gaming.

I'm very glad that development is continuing and consistent. Thank you!
